### ISSUE TITLE
[EKS]: EKS Cluster Tagging Propogation

### ISSUE THREAD

**Issue by tabern** [aws/containers-roadmap#374]
**Tell us about your request**
Ability to propogate EKS cluster tags to EC2 resources (instances, autoscaling groups, load balancers) used as part of the cluster. This will help with resource organization and cost allocation within your AWS account.

**Feedback requested**
Should we expand the scope of this to include propagating K8s labels to AWS tags and giving you the ability to assign labels used for resources within certain namespaces, services, etc as AWS tags on related resources? This is one path to improving cost allocation capabilities withing AWS and K8s.

**Comment by tdmalone** [comment]
> Should we expand the scope of this to include propagating K8s labels to AWS tags and giving you the ability to assign labels used for resources within certain namespaces, services, etc as AWS tags on related resources?

Yes - I think this is a great idea. It would be a really ace integration and, other than just for cost allocation, it would make it easier to programatically select k8s-managed resources in custom ops scripting.

**Comment by xanderbo** [comment]
First, make it work. Then make it work better. Expanded scope is great as long as it does not delay the release date.

**Comment by AlexHornDev** [comment]
Any ETA for this ?

**Comment by gmagella-ca** [comment]
I was just deploying a ALB using ALB Ingress controller and I notice that we do have the ability to tag it by annotating the Ingress. 

  annotations:
    alb.ingress.kubernetes.io/tags: auto-delete=no,owner=gmagella,app=my-1st-flask

Looks like it does tag the ALB and the TG. But I couldn't help but notice that it didn't update tags on the ALB, when changing the Ingress details. (it does update the TG tags though). 

Let me know if this is not the right venue to comment on this, and I'll take appropriate actions.

**Comment by i5okie** [comment]
Would be nice to propagate these tags to EBS volumes too.

**Comment by cdenneen** [comment]
Did the tags used to work? I thought they did but all my nodeGroups now no longer have the tags specified in the eksctl configuration. Need to get these tags back in as they are used for Cost reporting.

**Comment by tabern** [comment]
eksctl can propogate tags for self-managed node groups, however tags for EKS managed node groups do not propogate. We are working on enabling tag propogation for EKS clusters. From the cluster down to nodes.

**Comment by jasongaunt** [comment]
Hi @tabern is there anymore news on this please? ♥

**Comment by MichaelX99** [comment]
I'm not exactly sure which issue to put this on so I'll probably but it in a decent number of issues but not having propagating Tags to all EKS cluster resources effectively makes this entire service useless.  If you can't tag your resources you can't separate out costs between different programs and therefore will never be allowed by a program manager and rightfully so.  This needs to be fixed ASAP.

**Comment by ayanich** [comment]
it would be great if labels on EKS resources could be used as cost-allocation tags. The few taggable EKS resources currently do minimal to help us with cost allocation.  We need tagging and propagation.

**Comment by Andrey9kin** [comment]
Would love to have tags propagated for EKS created security group

**Comment by tuxtek** [comment]
I agree with the above statement around making it work as the priority. With the heavy dependency and enforcement around tagging, it is upsetting to see that the behavior we have come to expect with tag propagation is not present in EKS. This is causing us to have to create exceptions and workarounds to problems such as auto scaling groups not being tagged and slowing down velocity as we look to migrate from ECS to EKS.

It would be great to understand what sort of time line we are looking at for resolving this as the priority.

Outside of fixing what is missing today, enabling additional k8s label propagation to aid in further dissecting costs is secondary, I would think. For us, fixing the tag propagation actually gives all of the data we need for cost allocation in EKS.

**Comment by saleem-mirza** [comment]
`eksctl` 0.48 will be supporting user defined tagging for EKS resources. Please see https://github.com/weaveworks/eksctl/releases/tag/0.48.0-rc.0

**Comment by nikitacr7** [comment]
+1. It'll be very good to propagate user defined tags to ENIs using Launch Template

**Comment by TBBle** [comment]
For instances specifically, you can set the tags in the Launch Template instead of setting them somewhere higher and propagating them down.

As well as being the approach that [AWS has already implemented and recommends for this use-case](https://github.com/aws/containers-roadmap/issues/608#issuecomment-675101463), this mechanism has the advantage that you can propagate those or other tags to associated EBS volumes for the instance, which ASG-propagated tags do not support. See [Tag Auto Scaling groups and instances](https://docs.aws.amazon.com/autoscaling/ec2/userguide/autoscaling-tagging.html) and [Creating a launch template for an Auto Scaling group](https://docs.aws.amazon.com/autoscaling/ec2/userguide/create-launch-template.html) for reference.

**Comment by saleem-mirza** [comment]
eksctl supports custom tag propagation to underlaying resources

**Comment by TBBle** [comment]
I don't think eksctl supports any _propagation_ beyond what EKS already supports. It does have some features to _copy_ tags around from place to place, mostly related to what I described in my previous comment, and for working around #608.

That doesn't cover the use cases described in _this_ ticket, such as "Setting a tag `CostCentre=Project2` on the EKS Cluster object will automatically put that tag on the NLB instance created for a LoadBalancer `Service`".

**Comment by saleem-mirza** [comment]
@TBBle `eksctl` do support tag propagation since version 0.48. However, tags are only propagated to ec2 and ebs volumes.

`ALB` or `ELB` are not created when `EKS` cluster is created. However, you can always use annotations to push tags to load balancers. see https://kubernetes-sigs.github.io/aws-load-balancer-controller/v2.2/guide/service/annotations/

**Comment by TBBle** [comment]
That was my point. The only tag _propagation_ eksctl supports is the one EKS implements: from ASGs to their EC2 instances and from Launch Templates to their EC2 instances and EBS instances. That's the PR you linked to, which implements the eksctl side of that EKS feature. It doesn't do anything more for this than can be done with e.g., CDK or terraform, because it calls the same APIs.

_This ticket_ is asking for exactly the propagation you've noted, for things that do _not_ exist when the cluster is created, but are created later, and still need to be visible for, e.g. cost allocation. Because eksctl is not running "later", it can't support that without an EKS feature behind it. Which is what this ticket is effectively requesting.

The _workaround_ for NLB for example is annotations on the Load Balancer, but that's not propagation, that's just repeating the same tags all through your config (and other cluster user's configs).

**Comment by TechIsCool** [comment]
I just encountered another location that tag propagation would be helpful. Granting access via Tags in AWS CloudWatch. the `/aws/eks/*/clusters:*` audit logs don't have any tags attached from the eks cluster.

**Comment by andylim0221** [comment]
+1. It'd be useful especially we are working from Terraform environment.

**Comment by mikestef9** [comment]
While EKS does not support custom tag propagation which this issue covers, a noteworthy recent improvement we have made is adding a cluster name AWS cost allocation tag to every EC2 instance that joins an EKS cluster. Details in the What's New [post](https://aws.amazon.com/about-aws/whats-new/2022/08/amazon-eks-cluster-level-cost-allocation-tagging/).

**Comment by vce-xx** [comment]
What we are really expecting is to be able to work with our own tagging system already in place.

**Comment by mamoit** [comment]
> [edit] eksctl can propagate tags for self-managed and managed node groups. however if you use CFN or the API, tags for EKS managed node groups do not propagate. We are working on enabling tag propagation for EKS clusters from the cluster down to nodes through the EKS API.

@tabern 2 years have passed, still working on it, or should this issue be moved to another column on the roadmap?

IMHO the "use eksctl" approach is not a valid one since if you're doing things in a repeatable way by using `terraform`, or the likes of it (anything that relies on the API) you won't be using `eksctl`.
This issue messes up our billing breakdown pretty severely since it relies on tags and therefore there is a lot of unaccounted for resources in the end of the month.

**Comment by praveenperera** [comment]
This basic feature is missing, can't even forecast costs properly without it

**Comment by slam-harmonicinc** [comment]
@tabern We are still waiting for this feature. Any ETA?

**Comment by andrewegel** [comment]
We would like to see this implemented as well, specifically for the Cloud Watch Log Groups

**Comment by seyal84** [comment]
Any update on this ?

**Comment by DhruvBundheliya** [comment]
Still waiting for this.

**Comment by david-w-webb** [comment]
I started a PoC to define a standard CloudFormation template to create private subnets and EKS cluster.  Auto-Mode was released in December, and looked like the perfect way to apply Karpenter without needing to configure it independently using Kubernetes yamls.

Alas, Auto-Mode doesn't apply the parent cluster's tags to its nodes.  The initial recommendation from AWS Support was to apply an AWS::EKS::NodeClass resource to the cluster, and include the tags in the NodeClass.  However, NodeClass isn't implemented in CloudFormation as-of this writing.

Preventing and eliminating _Troublesome Untagged Resources Deployed Spontaneously_ **(TURDS)** has become a bit part of my existence over the past few years.

I'm reverting back to a managed NodeGroup (which can be tagged). and working on implementing CNI and Karpenter directly via CloudFormation instead of using Auto-Mode until I can propagate tags (and separate Node subnets from Pod subnets within Auto-Mode).

**Comment by orolega** [comment]
I worked around this by using the below to patch the NodeClass. I also set `dataplane_wait_duration = "60s"` in the `"terraform-aws-modules/eks/aws"` module, though unsure if that was really needed. Regardless, end result was the nodeclass was patched with the tags before resources were created so the tags were there when they were created.

```
resource "kubectl_manifest" "nodeclass_tags" {
  yaml_body = yamlencode({
    apiVersion = "eks.amazonaws.com/v1"
    kind       = "NodeClass"
    metadata   = { name = "default" }
    spec       = { tags = var.tags }
  })

  server_side_apply = true
  force_conflicts   = true
}
```

**Comment by CC-Unit-IT** [comment]
> I worked around this by using the below to patch the NodeClass. I also set `dataplane_wait_duration = "60s"` in the `"terraform-aws-modules/eks/aws"` module, though unsure if that was really needed. Regardless, end result was the nodeclass was patched with the tags before resources were created so the tags were there when they were created.
> 
> ```
> resource "kubectl_manifest" "nodeclass_tags" {
>   yaml_body = yamlencode({
>     apiVersion = "eks.amazonaws.com/v1"
>     kind       = "NodeClass"
>     metadata   = { name = "default" }
>     spec       = { tags = var.tags }
>   })
> 
>   server_side_apply = true
>   force_conflicts   = true
> }
> ```
@orolega 

And this patch persists for longer?

**Comment by orolega** [comment]
@CC-Unit-IT so far yes. Really don't see why this hasn't been implemented within the `"terraform-aws-modules/eks/aws"` module itself yet

**Comment by bryantbiggs** [comment]
We don't use the K8s providers such as Helm, kubectl, or Kubernetes, within the module. That just brings a lot of issues and challenges. It's an AWS feature request, not an IaC hack

**Comment by orolega** [comment]
@bryantbiggs I wasn't suggesting my workaround of using the kubectl provider to be implemented. I am saying that the "default" nodeclass created clearly supports tags as evident through my workaround. So I'm questioning why hasn't user supplied tags been able to be applied to the nodeclass on creation through the module. Basically the functionality exists at the resource level, so why isn't it exposed at the module level?

**Comment by cdenneen** [comment]
Module doesn't create the built-in NodePools just tells API to create those (like AWS Console does) so it's AWS that needs to support it via API but doesn't. I ended up with an AWS event and lambda for mine to tag the nodes because while your patch works I'm curious during updates if they would be removed since those are managed at CP level on the cluster (hence the managed-by eks on those NodePool and NodeClasses).

**Comment by bryantbiggs** [comment]
What @cdenneen said
